### PR TITLE
Xsd datatypes validation

### PR DESCRIPTION
**Pull request by josepajay**
This PR will close the issue https://github.com/GSS-Cogs/csv-validation/issues/54

### CODE REVIEW

File: src/test/scala/DateFormatTest.scala [hunk 1]
```diff
@@ -6,36 +6,243 @@ class DateFormatTest extends FunSuite {
...
+
+  }
+
+  test("it should parse date") {
+    val dateFormatObj =
+      DateFormat(None, "http://www.w3.org/2001/XMLSchema#date")
+    val maybeParsedDateTime = dateFormatObj.parse(
+      "-0045-01-01"
+    )
+    // edge case "-0045-01-01" is a valid date but it cant be parsed properly
```
Comment (robons) [inline src/test/scala/DateFormatTest.scala:230]: This comment seems a bit out of date if the test passes.

File: src/main/scala/TableGroup.scala [hunk 2]
```diff
@@ -423,4 +425,13 @@ case class TableGroup private (
       case (_, table) => table.warnings
     })
   warnings = warnings.concat(warningsFromTables)
+
+  def validateRow(row: CSVRecord, tableUrl: String) = {
+    // revisit if we need a variable validatedTables (hash) as in csvlint
+    tables(tableUrl).validateRow(row)
```
Comment (robons) [inline src/main/scala/TableGroup.scala:431]: Whilst the performance of looking objects up in a dictionary shouldn't be *too* bad, it's still better to only look them up once instead of one per row. 

Perhaps we could re-write the validator so that it calls `tableGroup.getTable(tableUrl)` and then directly calls `table.validateRow(row)` itself? That way it can keep a hold of which table it is and avoid having to look it up thousands of times.

File: src/test/resources/csvwExamples/observations_duplicate_headers.csv [hunk 3]
```diff
@@ -0,0 +1,10001 @@
+Geography,Age,Age,Health,Value
```
Comment (robons) [inline src/test/resources/csvwExamples/observations_duplicate_headers.csv:1]: Can we try and make some of these CSV files smaller?

File: src/main/scala/Validator.scala [hunk 4]
```diff
@@ -1,18 +1,103 @@
...
         Right(tableGroup.warnings.map(w => processWarnings(w)))
+      }
       case Left(errorMessage) => Left(errorMessage)
     }
   }
 
+  def fetchSchemaTables(schema: TableGroup) = {
+    for (tableUrl <- schema.tables.keys) {
+      val tableUri = new URI(tableUrl)
+      if (tableUri.getScheme == "file") { // tableUrl is a filename in this case, so check if it exists
```
Comment (robons) [inline src/main/scala/Validator.scala:38]: I'd move this file-not-exists check into `readCSV`. There's no point having two if statements and two `new File(...)` calls where one would do just fine.

File: src/main/scala/Validator.scala [hunk 5]
```diff
@@ -1,18 +1,103 @@
...
+      // Call validate csv function
+    }
+  }
+
+  def readCSV(tableUri: URI) = {
+    val parser = if (tableUri.getScheme == "file") {
+      CSVParser
+        .parse(
+          new File(tableUri),
+          StandardCharsets.UTF_8,
```
Comment (robons) [inline src/main/scala/Validator.scala:56]: Shouldn't we be using the `dialect.encoding` and `dialect.delimiter` information here to ensure we properly parse the CSV? 

(See the else statement below too)

File: src/main/scala/Validator.scala [hunk 6]
```diff
@@ -1,18 +1,103 @@
...
+          warnings :+= ErrorMessage(
+            "Blank rows",
+            "structure",
+            row.getRecordNumber.toString,
+            "",
+            "",
+            ""
+          )
+        }
+        schema match {
```
Comment (robons) [inline src/main/scala/Validator.scala:80]: This match statement should be outside of the for loop - I suspect the cost of it will add up quite quickly.

File: src/main/scala/Column.scala [hunk 7]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+    (value, format) =>
+      {
+        val validDecimalDatatypeRegex =
+          "(\\+|-)?([0-9]+(\\.[0-9]*)?|\\.[0-9]+)".r
+        if (!validDecimalDatatypeRegex.pattern.matcher(value).matches()) {
+          Left("invalid_decimal")
+        } else {
+          numericParser(value, format) match {
+            case Left(_)         => Left("invalid_decimal")
+            case Right(newValue) => Right(newValue.doubleValue())
```
Comment (robons) [inline src/main/scala/Column.scala:109]: I think we should be treating thia as a *Java* BigDecimal instead of as a double. I don't think the [XSD specification](https://www.w3.org/TR/xmlschema11-2/#decimal) limits the precision of decimals, so we will need to be able to cope with values that a double cannot represent.

**Review by robons — CHANGES_REQUESTED** [review]
There's a lot of good work here but a few changes we need to make before going forwards.

As ever, if you have any questions with what I'd said, ask away.

File: build.sbt [hunk 8]
```diff
@@ -28,7 +28,8 @@ libraryDependencies += "com.fasterxml.jackson.core" % "jackson-annotations" % "2
 libraryDependencies += "com.fasterxml.jackson.core" % "jackson-core" % "2.12.1"
 libraryDependencies += "com.softwaremill.sttp.client3" %% "core" % "3.0.0"
 libraryDependencies += "com.ibm.icu" % "icu4j" % "68.2"
-
+libraryDependencies += "org.apache.commons" % "commons-csv" % "1.8"
+libraryDependencies += "com.lihaoyi" %% "os-lib" % "0.7.7"
```
Comment (robons) [inline build.sbt:32]: Whats `os-lib` for?

File: src/main/scala/Column.scala [hunk 9]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+import scala.math.BigInt.javaBigInteger2bigInt
 
 object Column {
   val datatypeDefaultValue: ObjectNode = JsonNodeFactory.instance
     .objectNode()
     .set("@id", new TextNode("http://www.w3.org/2001/XMLSchema#string"))
 
+  val DatatypeParser
+      : Map[String, (String, Option[String]) => Either[String, Any]] = Map(
+    "http://www.w3.org/1999/02/22-rdf-syntax-ns#XMLLiteral" -> trimValue(),
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: okay, changed it

File: src/main/scala/Column.scala [hunk 10]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+    "http://www.w3.org/2001/XMLSchema#dateTimeStamp" -> processDateTimeStamp(),
+    "http://www.w3.org/2001/XMLSchema#gDay" -> processGDay(),
+    "http://www.w3.org/2001/XMLSchema#gMonth" -> processGMonth(),
+    "http://www.w3.org/2001/XMLSchema#gMonthDay" -> processGMonthDay(),
+    "http://www.w3.org/2001/XMLSchema#gYear" -> processGYear(),
+    "http://www.w3.org/2001/XMLSchema#gYearMonth" -> processGYearMonth(),
+    "http://www.w3.org/2001/XMLSchema#time" -> processTime()
+  )
+
+  def trimValue(): (String, Option[String]) => Either[String, String] =
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: 👍

File: src/main/scala/Column.scala [hunk 11]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
 import scala.collection.mutable.Map
 import scala.jdk.CollectionConverters.IteratorHasAsScala
+import scala.math.BigInt.javaBigInteger2bigInt
 
 object Column {
   val datatypeDefaultValue: ObjectNode = JsonNodeFactory.instance
     .objectNode()
     .set("@id", new TextNode("http://www.w3.org/2001/XMLSchema#string"))
 
+  val DatatypeParser
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: No, Changed it

File: src/main/scala/Column.scala [hunk 12]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+  def trimValue(): (String, Option[String]) => Either[String, String] =
+    (v, _) => Right(v.strip())
+
+  def allValueValid(): (String, Option[String]) => Either[String, String] =
+    (v, _) => Right(v)
+
+  def processBooleanDatatype()
+      : (String, Option[String]) => Either[String, Boolean] = { (v, format) =>
+    {
+      var formatValues = Array[String]()
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: Ah yes, changed it

File: build.sbt (hunk 8 above)
Comment (josepajay) [reply build.sbt:32]: This was added while experimenting with the paths issue we had, didn't remove after it. I got the code to successfully compile without it, so removing it for now.

File: src/main/scala/Column.scala [hunk 13]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+        }
+      }
+    }
+  }
+
+  def processDecimalDatatype()
+      : (String, Option[String]) => Either[String, Double] = {
+    (value, format) =>
+      {
+        val validDecimalDatatypeRegex =
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: Done

File: src/test/resources/csvwExamples/observations_duplicate_headers.csv (hunk 3 above)
Comment (josepajay) [reply src/test/resources/csvwExamples/observations_duplicate_headers.csv:1]: Done

File: src/test/scala/DateFormatTest.scala (hunk 1 above)
Comment (josepajay) [reply src/test/scala/DateFormatTest.scala]: Removed

File: src/main/scala/Column.scala [hunk 14]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+      : (String, Option[String]) => Either[String, BigInteger] = {
+    (value, format) =>
+      {
+        val validIntegerRegex = "[\\-+]?[0-9]+".r
+        if (!validIntegerRegex.pattern.matcher(value).matches()) {
+          Left("invalid_integer")
+        } else {
+          numericParser(value, format) match {
+            case Right(newValue) => {
+              try {
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: Done

File: src/main/scala/Column.scala [hunk 15]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+      }
+  }
+
+  def dateTimeParser(
+      datatype: String,
+      warning: String,
+      value: String,
+      format: Option[String]
+  ): Either[String, ZonedDateTime] = {
+    val dateFormatObject = if (format.isEmpty) {
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala:509]: True, not sure why I did this. May be some old code before format was not option? I have changed it

File: src/test/scala/ColumnTest.scala [hunk 16]
```diff
@@ -189,4 +189,524 @@ class ColumnTest extends FunSuite {
...
+    }
+  }
+
+  // Tests for processDoubleDatatype method
+  test("should process valid double datatype value without errors") {
+    val result = Column.processDoubleDatatype()("4268.22752E11", None)
+    assert(result.isRight)
+    result match {
+      case Right(doubleDatatype) => {
+        assert(doubleDatatype.toString == "4.26822752E14")
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/test/scala/ColumnTest.scala:224]: At many places the value is a bigInteger so direct comparison is not possible and thus the string conversion. Here the value is a double and conversion to string is not needed. I have changed it.

File: src/main/scala/Column.scala [hunk 17]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+
+  def processUnsignedLong()
+      : (String, Option[String]) => Either[String, BigInteger] = {
+    (value, format) =>
+      {
+        val result = processNonNegativeInteger()(value, format)
+        result match {
+          case Left(_) => Left("invalid_unsignedLong")
+          case Right(newValue) => {
+            if (newValue > Long.MaxValue) { // This is not quite right X2 this value is what I need. FIX ME
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: Done

File: src/main/scala/Column.scala [hunk 18]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+      : (String, Option[String]) => Either[String, Long] = { (value, format) =>
+    {
+      val validLongDatatypeRegex = "[\\-+]?[0-9]+".r
+      if (!validLongDatatypeRegex.pattern.matcher(value).matches()) {
+        Left("invalid_long")
+      } else {
+        processIntegerDatatype()(value, format) match {
+          case Left(warning) => Left(s"invalid_long - ${warning}")
+          case Right(newValue) => {
+            val doubleValue = newValue.doubleValue()
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: this variable was not being used. Removed it. These type of conversions was being used but then I switched to auto type conversions on comparisons provided by scala.math.BigInt.javaBigInteger2bigInt

File: src/main/scala/Column.scala [hunk 19]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+    (value, format) =>
+      {
+        val validIntegerRegex = "[\\-+]?[0-9]+".r
+        if (!validIntegerRegex.pattern.matcher(value).matches()) {
+          Left("invalid_integer")
+        } else {
+          numericParser(value, format) match {
+            case Right(newValue) => {
+              try {
+                val bigIntValue = newValue match {
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: done

File: src/test/scala/ColumnTest.scala [hunk 20]
```diff
@@ -189,4 +189,524 @@ class ColumnTest extends FunSuite {
...
+  }
+
+  test(
+    "should process valid decimal datatype value beginning with decimal point without errors - "
+  ) {
+    val result = Column.processDecimalDatatype()(".3", None)
+    assert(result.isRight)
+    result match {
+      case Right(value) => {
+        assert(value.toString == "0.3")
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/test/scala/ColumnTest.scala:632]: At many places the value is a bigInteger so direct comparison is not possible and thus the string conversion. Here the value is a double and conversion to string is not needed. I have changed it.

File: src/test/scala/ColumnTest.scala [hunk 21]
```diff
@@ -189,4 +189,524 @@ class ColumnTest extends FunSuite {
...
+    }
+  }
+
+  // Tests for processDecimalDatatype
+  test(
+    "should return error when invalid Decimal value is processed"
+  ) {
+    val result =
+      Column.processDecimalDatatype()(
+        "3,5", // commas are not permitted; the decimal separator must be a period
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/test/scala/ColumnTest.scala:612]: I have modified the comment to convey this.

File: src/main/scala/Table.scala [hunk 22]
```diff
@@ -498,4 +499,56 @@ case class Table private (
...
+          ""
+        )
+      } else (columnNames :+= columnName)
+      if (columnIndex < columns.length) {
+        val column = columns(columnIndex)
+        val WarningsAndErrors(w, e) = column.validateHeader(columnName)
+        warnings = warnings.concat(w)
+        errors = errors.concat(e)
+      } else {
+        errors :+= ErrorMessage("Malformed header", "Schema", "1", "", "", "")
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Table.scala:548]: This is when there are more columns in the first row of the CSV than the number of columns specified in the metadata. Not sure of this comment, especially the header definition part

File: src/main/scala/Column.scala [hunk 23]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
   val datatypeDefaultValue: ObjectNode = JsonNodeFactory.instance
     .objectNode()
     .set("@id", new TextNode("http://www.w3.org/2001/XMLSchema#string"))
 
+  val DatatypeParser
+      : Map[String, (String, Option[String]) => Either[String, Any]] = Map(
+    "http://www.w3.org/1999/02/22-rdf-syntax-ns#XMLLiteral" -> trimValue(),
+    "http://www.w3.org/1999/02/22-rdf-syntax-ns#HTML" -> trimValue(),
+    "http://www.w3.org/ns/csvw#JSON" -> trimValue(),
+    "http://www.w3.org/2001/XMLSchema#anyAtomicType" -> allValueValid(),
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: done

File: src/main/scala/PropertyChecker.scala [hunk 24]
```diff
@@ -792,10 +792,10 @@ object PropertyChecker {
...
             // In ruby regexp is stored in format key. Also regexp validated. Determine how to handle this in scala
             // value["format"] = Regexp.new(value["format"])
-            objectNode.set(
-              "format",
-              new TextNode(objectNode.get("format").asText)
-            )
+            objectNode
+              .get("format")
+              .asText
+              .r //Add a unit test for invalid format warning
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/PropertyChecker.scala]: Done

File: src/test/scala/DateFormatTest.scala [hunk 25]
```diff
@@ -6,36 +6,240 @@ class DateFormatTest extends FunSuite {
...
-      DateFormat(None, Some("http://www.w3.org/2001/XMLSchema#dateTime"))
-    assert(
-      dateFormatObj.getType() === "http://www.w3.org/2001/XMLSchema#dateTime"
+      DateFormat(None, "http://www.w3.org/2001/XMLSchema#gDay")
+    val maybeParsedDateTime = dateFormatObj.parse(
+      "---01-04:00"
+    )
+    assert(maybeParsedDateTime.isDefined)
+    val parsedDateTime = maybeParsedDateTime.get
+    assert(parsedDateTime.getDayOfMonth == 1)
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/test/scala/DateFormatTest.scala:93]: Done

File: src/test/scala/DateFormatTest.scala [hunk 26]
```diff
@@ -6,36 +6,240 @@ class DateFormatTest extends FunSuite {
...
+    assert(maybeParsedDateTime.isDefined)
+    val parsedDateTime = maybeParsedDateTime.get
+    assert(parsedDateTime.getYear == 2002)
+    assert(parsedDateTime.getMonthValue == 10)
+    assert(parsedDateTime.getDayOfMonth == 10)
+    assert(parsedDateTime.getHour == 12)
+    assert(parsedDateTime.getMinute == 0)
+    assert(parsedDateTime.getSecond == 0)
+    assert(parsedDateTime.getNano == 0.0123456 * 1e9)
   }
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/test/scala/DateFormatTest.scala:83]: Done

File: src/main/scala/Table.scala [hunk 27]
```diff
@@ -498,4 +499,56 @@ case class Table private (
     }
   }
   warnings = warnings.concat(warningsFromColumns)
+
+  def validateRow(row: CSVRecord): WarningsAndErrors = {
+    if (columns.nonEmpty) {
+      for ((value, i) <- row.iterator.asScalaArray.zipWithIndex) {
+        //catch any exception here, possibly outOfBounds  and set warning too many values
+        val column = columns(i)
+        column.validate(value, row.getRecordNumber)
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Table.scala:508]: Done

File: src/main/scala/Column.scala [hunk 28]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
   TextNode
 }
 
+import java.math.BigInteger
+import java.time.ZonedDateTime
 import scala.collection.mutable.Map
 import scala.jdk.CollectionConverters.IteratorHasAsScala
+import scala.math.BigInt.javaBigInteger2bigInt
 
 object Column {
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala:21]: @robons To review after latest changes

File: src/main/scala/Validator.scala (hunk 5 above)
Comment (josepajay) [reply src/main/scala/Validator.scala:52]: Leaving this until we work on dialects

File: src/main/scala/Column.scala [hunk 29]
```diff
@@ -9,14 +9,512 @@ import com.fasterxml.jackson.databind.node.{
...
+          }
+        }
+      }
+  }
+
+  def numericParser(
+      value: String,
+      format: Option[String]
+  ): Either[String, Number] = {
+    val numberFormatObject = NumberFormat(format, None, None)
```
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala]: Done

File: src/main/scala/Column.scala (hunk 28 above)
[the remark this answers is not in the document]
Comment (robons) [reply src/main/scala/Column.scala:28]: I do think it's too big, for now could you just move the Column *object* into a separate file. Call it `Column.object.scala` maybe?

File: src/main/scala/models/Format.scala [hunk 30]
```diff
@@ -0,0 +1,8 @@
+package CSVValidation
+
+case class Format(
+    maybeBooleanFormatOrRegExFormat: Option[String],
```
Comment (robons) [inline src/main/scala/models/Format.scala:4]: shouldn't this be stuck into `pattern` since that's where it comes from?

File: src/main/scala/Column.scala [hunk 31]
```diff
@@ -245,12 +278,47 @@ object Column {
...
+      errors = Array()
     )
   }
 
+  private def getMaybeFormatForColumn(formatNode: JsonNode) = {
+    if (formatNode.isMissingNode || formatNode.isNull) {
+      None
+    } else {
+      formatNode match {
+        case s: TextNode => Some(Format(Some(s.asText()), None, None, None))
```
Comment (robons) [inline src/main/scala/Column.scala:294]: Yeah, I think the format string should go into the `format` property instead of this new special location.

File: src/main/scala/Column.scala [hunk 32]
```diff
@@ -288,6 +366,708 @@ case class Column private (
...
+              "schema",
+              rowNumber.toString,
+              columnOrdinal.toString,
+              s"'$v' - ${errorMessageContent.content}",
+              datatype.toPrettyString
+            )
+          }
+          case Right(s) => {}
+        }
+//        val (value, warning) = //CREATE DATATYPE PARSERS
```
Comment (robons) [inline src/main/scala/Column.scala:1036]: Can we remove this now?

File: src/main/scala/Column.scala [hunk 33]
```diff
@@ -288,6 +366,708 @@ case class Column private (
...
+    if (nullParam.contains(value)) {
+      addErrorIfRequiredValueAndValueEmpty(value, rowNumber)
+    } else {
+      val values = separator match {
+        case Some(separator) => value.split(separator)
+        case None            => Array[String](value)
+      }
+      val datatypeBaseField = datatype.path("base")
+      val datatypeOfColumn = if (datatypeBaseField.isMissingNode) {
+        datatype.get("@id").asText()
```
Comment (robons) [inline src/main/scala/Column.scala:1017]: Again, it feels weird to be doing JSON parsing in here, can we add a property to the column which holds the baseDataType? And then populate that in the fromJson method?

File: src/main/scala/Column.scala [hunk 34]
```diff
@@ -288,6 +366,708 @@ case class Column private (
...
+            )
+          }
+          case Right(s) => {}
+        }
+//        val (value, warning) = //CREATE DATATYPE PARSERS
+//        val format = ""
+//        val (value, warning) = Column.DatatypeParser(s)(value, format)
+      }
+    }
+    WarningsAndErrors(Array(), Array())
```
Comment (robons) [inline src/main/scala/Column.scala:1041]: Do we not want to pass the errors and warnings that we've generated up?

File: src/main/scala/Validator.scala [hunk 35]
```diff
@@ -1,19 +1,98 @@
...
+import java.net.URI
+import java.nio.charset.StandardCharsets
+import scala.jdk.CollectionConverters.IterableHasAsScala
+
 // This is only a skeleton for the core Validator class to be created. It is planned to accept the source csv and
 // schema and return errors and warnings collection when validate method is invoked. Changes to this signature can be
 // made in the due course or as per requirements
-class Validator(val filePath: String) {
+class Validator(var tableCsvFile: URI, sourceUri: String = "") {
+  val isCsvHeader = true
```
Comment (robons) [inline src/main/scala/Validator.scala:14]: is this used?

File: src/main/scala/Column.scala [hunk 36]
```diff
@@ -288,6 +366,708 @@ case class Column private (
...
+  }
+
+  def validateHeader(columnName: String): WarningsAndErrors = {
+    var errors = Array[ErrorWithCsvContext]()
+    titles match {
+      case Some(titles) => {
+        var validHeaders = Array[String]()
+        for ((_, v) <- titles.asInstanceOf[ObjectNode].getKeysAndValues) {
+          val titlesArray = Array.from(v.elements().asScala)
```
Comment (robons) [inline src/main/scala/Column.scala:1050]: Again, since this is JSON parsing, it'd be nicer if we could have done this in the `fromJson` function.

Comment (josepajay) [reply src/main/scala/Column.scala]: Ya, definitely agree with that. Changed

File: src/main/scala/Validator.scala (hunk 35 above)
Comment (josepajay) [reply src/main/scala/Validator.scala]: No it wasn't. Changed

File: src/main/scala/Column.scala (hunk 34 above)
Comment (josepajay) [reply src/main/scala/Column.scala]: Yes they should be pass back up, but this function is not complete and thus codes like this. These are problems associated with breaking up a large feature to smaller PRs.  For time being I have returned the errors and warnings I have.

File: src/main/scala/Column.scala (hunk 32 above)
Comment (josepajay) [reply src/main/scala/Column.scala]: Removed

File: src/main/scala/Column.scala (hunk 31 above)
Comment (josepajay) [reply src/main/scala/Column.scala]: Okay, reusing pattern property now.

File: src/main/scala/models/Format.scala (hunk 30 above)
Comment (josepajay) [reply src/main/scala/models/Format.scala]: Done

File: src/main/scala/Column.scala (hunk 33 above)
Comment (josepajay) [reply src/main/scala/Column.scala]: Done

File: src/main/scala/Column.scala (hunk 28 above)
[the remark this answers is not in the document]
Comment (josepajay) [reply src/main/scala/Column.scala:28]: Leaving this for now, since companion object must be in the same file.

**Review by robons — APPROVED** [review]
